### PR TITLE
POT Accounting in CAFs using BNB Retriever module

### PR DESCRIPTION
**Pull request by jzettle**
Adding POT Accounting to CAFs usign output of BNBRetriever module. Fills in SRHeader with one entry per event. More ideal is an entry per subrun.

### CODE REVIEW

File: sbncode/CAFMaker/CAFMaker_module.cc [hunk 1]
```diff
@@ -443,12 +445,56 @@ void CAFMaker::beginRun(art::Run& run) {
...
   // get POT information
-  art::Handle<sumdata::POTSummary> pot_handle;
-  sr.getByLabel("generator", pot_handle);
-
-  if (pot_handle.isValid()) {
-    fSubRunPOT = pot_handle->totgoodpot;
-    fTotalPOT += fSubRunPOT;
+  fIsRealData = true;
+  if(fIsRealData)
+  {
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/CAFMaker_module.cc:450]: Can we pull this out into a separate `FillBNBPOTInfo` function? Maybe even in a new `FillExposure.cxx`? Hopefully we will soon also have a NuMI version of this, which is going to make beginSubRun very long.

File: sbncode/CAFMaker/CAFMaker_module.cc [hunk 2]
```diff
@@ -1368,6 +1414,8 @@ void CAFMaker::produce(art::Event& evt) noexcept {
   rec.hdr.det     = fDet;
   rec.hdr.fno     = fFileNumber;
   rec.hdr.pot     = fSubRunPOT;
+  if(fIsRealData)
+    rec.hdr.bnbinfo = bnb_info_store;
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/CAFMaker_module.cc:1418]: This is duplicating all the bnbinfo into every event, right? I think we probably only want to write it once. Not sure if you should put an `&& fFirstInFile` here, or just a `.clear()` after you've copied it into the record.

File: sbncode/CAFMaker/CAFMaker_module.cc [hunk 3]
```diff
@@ -167,6 +168,7 @@ class CAFMaker : public art::EDProducer {
   double fSubRunPOT;
   double fTotalSinglePOT;
   double fTotalEvents;
+  std::vector<caf::SRBNBInfo> bnb_info_store;
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/CAFMaker_module.cc:171]: Can we give this a name following the convention? `fBNBInfo` is probably fine. And a comment like `///< Store detailed BNB info to save into the first StandardRecord of the output file`

**Review by cjbackhouse — CHANGES_REQUESTED** [review]
Looks good, but I have a bunch of small comments about reorganizing this.

File: sbncode/CAFMaker/CAFMaker_module.cc [hunk 4]
```diff
@@ -443,12 +445,56 @@ void CAFMaker::beginRun(art::Run& run) {
 void CAFMaker::beginSubRun(art::SubRun& sr) {
   // get the POT
   // get POT information
-  art::Handle<sumdata::POTSummary> pot_handle;
-  sr.getByLabel("generator", pot_handle);
-
-  if (pot_handle.isValid()) {
-    fSubRunPOT = pot_handle->totgoodpot;
-    fTotalPOT += fSubRunPOT;
+  fIsRealData = true;
```
[the remark this answers is not in the document]
Comment (jzettle) [reply sbncode/CAFMaker/CAFMaker_module.cc:448]: I'm sorry I did this in somewhat in a rush and forgot to clean this up. I added this to test things as I thought this was set somewhere when I first wrote this in but I found it wasn't. I will fix this by removing that forced set.

File: sbncode/CAFMaker/CAFMaker_module.cc (hunk 1 above)
Comment (jzettle) [reply sbncode/CAFMaker/CAFMaker_module.cc:450]: Sure, that was actually my first thought. I will work towards that

File: sbncode/CAFMaker/CAFMaker_module.cc (hunk 2 above)
Comment (jzettle) [reply sbncode/CAFMaker/CAFMaker_module.cc:1418]: Yes, this is what we discussed before. I pushed this before resolving that. Will add that in too. I can try a `fFirstInFile` and see if that works that's really what we want

File: sbncode/CAFMaker/CAFMaker_module.cc (hunk 3 above)
Comment (jzettle) [reply sbncode/CAFMaker/CAFMaker_module.cc:171]: Yep, my bad I forgot to follow the convention here

File: sbncode/CAFMaker/CAFMaker_module.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (cjbackhouse) [reply sbncode/CAFMaker/CAFMaker_module.cc:448]: Ah, the other PR is not merged yet, but you'll want to make sure you don't clash too badly with it https://github.com/SBNSoftware/sbncode/pull/209

**Comment by jzettle** [comment]
Okay, update in latest commit that I think addresses all comments from @cjbackhouse. Tested and works as I think intended where there's only one non-zero entry at the firstInFile. Thanks for the review and any other comments I'm happy to address.

File: sbncode/CAFMaker/CAFMaker_module.cc [hunk 5]
```diff
@@ -443,12 +446,28 @@ void CAFMaker::beginRun(art::Run& run) {
...
-    fTotalPOT += fSubRunPOT;
+  if(fIsRealData)
+  {
+    fBNBInfo.clear();
+    fSubRunPOT = 0;
+    art::Handle<std::vector<sbn::BNBSpillInfo> > bnb_spill;
+    sr.getByLabel(fParams.POTDataLabel(), bnb_spill);
+    std::vector<art::Ptr<sbn::BNBSpillInfo> >  bnb_spill_info;
+    if (bnb_spill.isValid()) {
+      art::fill_ptr_vector(bnb_spill_info, bnb_spill);
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/CAFMaker_module.cc:457]:
```suggestion
    if (bnb_spill.isValid()) {
      std::vector<art::Ptr<sbn::BNBSpillInfo> >  bnb_spill_info;
      art::fill_ptr_vector(bnb_spill_info, bnb_spill);
```

File: sbncode/CAFMaker/CAFMaker_module.cc [hunk 6]
```diff
@@ -1385,6 +1409,8 @@ void CAFMaker::produce(art::Event& evt) noexcept {
   StandardRecord* prec = &rec;
   fRecTree->SetBranchAddress("rec", &prec);
   fRecTree->Fill();
+  fBNBInfo.clear();
+  rec.hdr.pot = 0;
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/CAFMaker_module.cc:1413]: Can you move these down two lines, below the srcol stuff? Otherwise you are wiping out the POT information from the version of the StandardRecord that gets saved into the art file. We don't use that for much/anything yet, but that will be an unpleasant surprise one day.

File: sbncode/CAFMaker/FillExposure.cxx [hunk 7]
```diff
@@ -0,0 +1,39 @@
+
+#include "FillExposure.h"
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/FillExposure.cxx:2]:
```suggestion
#include "sbncode/CAFMaker/FillExposure.h"
```

File: sbncode/CAFMaker/FillExposure.h [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Principal/Handle.h"
+#include "art/Framework/Principal/SubRun.h"
+
+#include "sbnanaobj/StandardRecord/StandardRecord.h"
+#include "sbnanaobj/StandardRecord/SRHeader.h"
+#include "sbnanaobj/StandardRecord/SRBNBInfo.h"
+#include "sbnobj/Common/POTAccounting/BNBSpillInfo.h"
+#include <vector>
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/FillExposure.h:15]: The majority of these includes are unnecessary here.

File: sbncode/CAFMaker/CAFMakerParams.h [hunk 9]
```diff
@@ -60,6 +60,12 @@ namespace caf
       Comment("List of suffixes to add to TPC reco tag names (e.g. cryo0 cryo1)")
     };
 
+    Atom<string> POTDataLabel {
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/CAFMakerParams.h:63]: Can we call this `BNBPOTInfoLabel` or similar, to allow ourselves room for a `NuMIPOTInfoLabel` in future?

**Review by cjbackhouse — APPROVED** [review]
Looking good. A few minor nits.

**Comment by jzettle** [comment]
Those comments resolved as well. Still working as before

File: sbncode/CAFMaker/CAFMaker_module.cc [hunk 10]
```diff
@@ -443,12 +446,28 @@ void CAFMaker::beginRun(art::Run& run) {
 void CAFMaker::beginSubRun(art::SubRun& sr) {
   // get the POT
   // get POT information
-  art::Handle<sumdata::POTSummary> pot_handle;
-  sr.getByLabel("generator", pot_handle);
-
-  if (pot_handle.isValid()) {
-    fSubRunPOT = pot_handle->totgoodpot;
-    fTotalPOT += fSubRunPOT;
+  if(fIsRealData)
```
Comment (PetrilloAtWork) [inline sbncode/CAFMaker/CAFMaker_module.cc:449]: We don't know this yet.

File: sbncode/CAFMaker/CAFMaker_module.cc [hunk 11]
```diff
@@ -443,12 +446,28 @@ void CAFMaker::beginRun(art::Run& run) {
...
+      FillExposure(bnb_spill_info, fBNBInfo, fSubRunPOT);
+      fTotalPOT += fSubRunPOT;
+    }
+  }
+  else
+  {
+    art::Handle<sumdata::POTSummary> pot_handle;
+    sr.getByLabel(fParams.GenLabel(), pot_handle);
+    
+    if (pot_handle.isValid()) {
```
Comment (PetrilloAtWork) [inline sbncode/CAFMaker/CAFMaker_module.cc:467]: Fun with C++ (not particularly endorsing this change):
```suggestion
    if (auto pot_handle = sr.getHandle<sumdata::POTSummary>(fParams.GenLabel())) {
```

File: sbncode/CAFMaker/CAFMaker_module.cc (hunk 10 above)
Comment (PetrilloAtWork) [reply sbncode/CAFMaker/CAFMaker_module.cc:449]: I suggest a duck-typing work-around (which still sucks because it may be NuMI instead of BNB etc.):
```suggestion
  auto const bnb_spill = sr.getHandle<std::vector<sbn::BNBSpillInfo>>(fParams.BNBPOTDataLabel(), bnb_spill);
  fIsRealData = bnb_spill.isValid();
  if(fIsRealData)
```
(of course then you don't need to retrieve `bnb_spill` again in the `if` branch).

Unfortunately, for some mysterious reason `art::SubRun::isRealData()` is not exposed (thus, I have no idea if it is correctly set, either).

**Review by PetrilloAtWork — CHANGES_REQUESTED** [review]
There is a problem in the logic of `beginSubRun()` that needs to be addressed.
For the rest, C++ suggestions here and there.

File: sbncode/CAFMaker/CAFMaker_module.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (jzettle) [reply sbncode/CAFMaker/CAFMaker_module.cc]: I reviewed that PR after the fact and I think everything is fine. As long as fIsRealData gets set via the fcl before beginSubRun is called then it should work. It looks like that is the case. Would want to confirm after this happens. I've made sure to remove the hardcode I used for testing

File: sbncode/CAFMaker/CAFMaker_module.cc (hunk 10 above)
Comment (jzettle) [reply sbncode/CAFMaker/CAFMaker_module.cc:449]: There's a PR that Chris mentioned #209 that puts this in the fcl so I'm inclined to let it be set there? If not we'll probably want to coordinate that. I'm not sure where else this fIsRealData value may be used in the future.

File: sbncode/CAFMaker/CAFMaker_module.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (brucehoward-physics) [reply sbncode/CAFMaker/CAFMaker_module.cc]: Heads up: That PR no longer uses a fhicl parameter - it was changed yesterday to determine it from the `art` data memeber (`evt.isRealData()`). *It is no longer set by beginSubrun but in the process step itself* 😬

File: sbncode/CAFMaker/FillExposure.cxx (hunk 7 above)
Comment (PetrilloAtWork) [reply sbncode/CAFMaker/FillExposure.cxx]: Idle comment about this: I would argue that `FillExposure.h` and `FillExposure.cxx` living in the same directory is a requirement. If that's true, isn't actually a better choice to go Bruce's way?

Currently, we use to officially recommend Chris's approach anyway.

File: sbncode/CAFMaker/CAFMaker_module.cc (hunk 10 above)
Comment (brucehoward-physics) [reply sbncode/CAFMaker/CAFMaker_module.cc:449]: (commented elsewhere in this PR just a moment ago that this is not the case any more after discussion at yesterdays SBN Ana Framework meeting. Note I'm not really following this PR that much but see my PR getting mentioned so wanted to clarify)

Comment (jzettle) [reply sbncode/CAFMaker/CAFMaker_module.cc:449]: Never mind, apparently this is not the case anymore (wow this stuff is hard to follow sometimes...). This is likely a good way to do this in light of `fIsRealData` being set in the process step so I need to set it in the subRun elsewhere to make this work for POT accounting...everything else should be self-consistent even with that change

File: sbncode/CAFMaker/FillExposure.cxx (hunk 7 above)
Comment (jzettle) [reply sbncode/CAFMaker/FillExposure.cxx]: I'm not sure what Bruce is doing, but I just followed the convention I see with all the other "Fill" files. I agree Chris's approach seems fine too.

File: sbncode/CAFMaker/CAFMaker_module.cc (hunk 10 above)
Comment (PetrilloAtWork) [reply sbncode/CAFMaker/CAFMaker_module.cc:449]: Actually in that pull request I was suggesting that `fIsRealData` be _removed_, exactly to avoid bugs like this.
The workaround I suggest here may affect that too:
1. we can decide to leave `fIsRealData` data member be, but then the use in `beginRun()` will suffer the same issue as here
2. we can decide to drop it, and then `fIsRealData` here becomes a local `isRealData` variable.

Comment (brucehoward-physics) [reply sbncode/CAFMaker/CAFMaker_module.cc:449]: I haven't fully looked at the code (or indeed most of the comments in this PR review outside of the few items I've been following that involved my PR 😅 ), but just throwing in my 2 cents with my current thoughts:

1) `fIsRealData` doesn't seem be being used in `beginRun()` at the moment though? Is that going to change with other suggestions or just a general comment? Unless you (@PetrilloAtWork ) are maybe considering how one might use such a data member consistently throughout the module (e.g. later when I have things in the `produce` step) or something?

2) Why not just get rid of a variable altogether (unless the answer to point 1 demands it) and just do `if ( bnb_spill.isValid() )`?

3) Could you then also do something similar for NuMI and do `if ( bnb_spill.isValid() || numi_spill.isValid() ) `

Comment (PetrilloAtWork) [reply sbncode/CAFMaker/CAFMaker_module.cc:449]: > fIsRealData doesn't seem be being used in beginRun() at the moment though? Is that going to change with other suggestions or just a general comment?

General comment. Until one week ago, `fIsRealData` was not used even in `beginSubRun()` and we were happy.
But then we forget the quirks, and six months from now we start using it, and much later we realize that we are missing information for one file per job (the issue is only for the first call of `begin[Sub]Run()`, after which `analyze()` kicks in; and those methods are in our situation called on each new input file).

Comment (cjbackhouse) [reply sbncode/CAFMaker/CAFMaker_module.cc:449]: > Unfortunately, for some mysterious reason art::SubRun::isRealData() is not exposed

I made an art issue about this years ago, and it was closed WONTFIX. They say that in principle you could mix data and MC events into a single file. Not sure how this is consistent with including it on Event given the possibility of overlays...

File: sbncode/CAFMaker/FillExposure.cxx (hunk 7 above)
Comment (cjbackhouse) [reply sbncode/CAFMaker/FillExposure.cxx]: All of our code uses full paths, except for a few places that don't (lots of places in StandardRecord :cry:). If I was in charge I would require full paths everywhere. Been bitten too many times by things being picked up from unexpected places.

**Comment by jzettle** [comment]
I think I've addressed everything.

File: sbncode/CAFMaker/FillExposure.cxx [hunk 12]
```diff
@@ -3,37 +3,43 @@
...
-        single_store.BTJT2 = info->BTJT2;
-        single_store.THCURR = info->THCURR;
-        single_store.M875BB = info->M875BB;
-        single_store.M876BB = info->M876BB;
-        single_store.MMBTBB = info->MMBTBB;
-        single_store.M875BB_spill_time_diff = info->M875BB_spill_time_diff;
-	single_store.M876BB_spill_time_diff = info->M876BB_spill_time_diff;
-        single_store.MMBTBB_spill_time_diff = info->MMBTBB_spill_time_diff;
-        BNBInfo.push_back(single_store);
+	subRunPOT += info.TOR875;
```
Comment (PetrilloAtWork) [inline sbncode/CAFMaker/FillExposure.cxx:41]: I assume that you considered and declined the suggestion of making this
```suggestion
	subRunPOT += info.POT();
```
which is fine.

**Review by PetrilloAtWork — APPROVED** [review]
I confirm all the necessary and some of the optional suggestions have been implemented.
I have left one comment on [one suggestion that hasn't been received](https://github.com/SBNSoftware/sbncode/pull/211#discussion_r762234436), which is fine but I don't know whether it was a deliberate choice or just missed do I pedantically point that out.
Approved regardless of how that part goes.

File: sbncode/CAFMaker/FillExposure.cxx (hunk 12 above)
Comment (jzettle) [reply sbncode/CAFMaker/FillExposure.cxx:41]: I did make that change, then I tried to put everything in a separate function and forgot to keep that change once I figured everything else out. Have recovered that change again.

**Comment by jzettle** [comment]
Comments addressed
